File: R/ifm_cash_rate_chart.py
"""
Replicate IFM Investors' "RBA Cash rate expectations" chart.

Plots the actual RBA cash rate as a dark step-function line, overlays
successive ~18-month futures strips as light grey/taupe lines, and
highlights the latest futures strip in orange with the endpoint annotated.
"""

import warnings
warnings.filterwarnings("ignore")

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import matplotlib.dates as mdates
import matplotlib.ticker as mticker
from datetime import datetime, timedelta
import pyreadr

# =============================================================================
# 1) Load futures data
# =============================================================================
rds = pyreadr.read_r("combined_data/all_data.Rds")
futures = list(rds.values())[0]

futures["date"] = pd.to_datetime(futures["date"])
futures["scrape_date"] = pd.to_datetime(futures["scrape_date"])

# Keep one observation per (scrape_date, contract expiry month)
futures = (
    futures.sort_values("scrape_date")
    .groupby(["scrape_date", "date"])
    .last()
    .reset_index()
)

# =============================================================================
# 2) Historical actual RBA cash rate (target rate changes since 2015)
# =============================================================================
# Source: RBA official target rate changes
rate_changes = [
    ("2015-02-04", 2.25),
    ("2015-05-06", 2.00),
    ("2016-05-04", 1.75),
    ("2016-08-03", 1.50),
    ("2019-06-05", 1.25),
    ("2019-07-03", 1.00),
    ("2019-10-02", 0.75),
    ("2020-03-04", 0.50),
    ("2020-03-20", 0.25),
    ("2020-11-04", 0.10),
    ("2022-05-04", 0.35),
    ("2022-06-08", 0.85),
    ("2022-07-06", 1.35),
    ("2022-08-03", 1.85),
    ("2022-09-07", 2.35),
    ("2022-10-05", 2.60),
    ("2022-11-02", 2.85),
    ("2022-12-07", 3.10),
    ("2023-02-08", 3.35),
    ("2023-03-08", 3.60),
    ("2023-05-03", 3.85),
    ("2023-06-07", 4.10),
    ("2023-08-08", 4.10),  # hold
    ("2023-11-08", 4.35),
    ("2025-02-18", 4.10),
    ("2025-04-01", 3.85),  # latest known
]

actual = pd.DataFrame(rate_changes, columns=["date", "rate"])
actual["date"] = pd.to_datetime(actual["date"])

# Extend to today so the step line continues
last_date = pd.Timestamp("2026-03-24")
actual = pd.concat(
    [actual, pd.DataFrame({"date": [last_date], "rate": [actual["rate"].iloc[-1]]})],
    ignore_index=True,
)

# =============================================================================
# 3) Sample futures strips — one per month (mid-month preference)
# =============================================================================
scrape_dates_sorted = sorted(futures["scrape_date"].unique())

# Group scrape dates by year-month, pick one near mid-month
scrape_df = pd.DataFrame({"scrape_date": scrape_dates_sorted})
scrape_df["ym"] = scrape_df["scrape_date"].values.astype("datetime64[M]")
scrape_df["day"] = scrape_df["scrape_date"].dt.day

sampled_scrapes = (
    scrape_df.assign(dist=lambda d: (d["day"] - 15).abs())
    .sort_values("dist")
    .groupby("ym")
    .first()
    .reset_index()["scrape_date"]
    .tolist()
)

# Also always include the latest scrape
latest_scrape = max(scrape_dates_sorted)
if latest_scrape not in sampled_scrapes:
    sampled_scrapes.append(latest_scrape)

# =============================================================================
# 4) Build the plot
# =============================================================================
fig, ax = plt.subplots(figsize=(10, 6.5))

# Background colour — white
fig.patch.set_facecolor("white")
ax.set_facecolor("white")

# --- Plot old futures strips (grey/taupe) ---
for sd in sorted(sampled_scrapes):
    strip = futures[futures["scrape_date"] == sd].sort_values("date")
    if strip.empty:
        continue
    # Only keep ~18 months forward from scrape date
    strip = strip[strip["date"] <= sd + pd.DateOffset(months=18)]
    if len(strip) < 2:
        continue
    if sd == latest_scrape:
        continue  # plot latest separately
    ax.plot(
        strip["date"],
        strip["cash_rate"],
        color="#b5a99a",   # taupe/grey matching IFM style
        linewidth=0.7,
        alpha=0.55,
        zorder=1,
    )

# --- Plot latest futures strip (orange) ---
latest_strip = futures[futures["scrape_date"] == latest_scrape].sort_values("date")
latest_strip = latest_strip[
    latest_strip["date"] <= latest_scrape + pd.DateOffset(months=18)
]
if not latest_strip.empty:
    ax.plot(
        latest_strip["date"],
        latest_strip["cash_rate"],
        color="#e8731a",  # IFM orange
        linewidth=3,
        zorder=3,
        solid_capstyle="round",
    )
    # Annotate endpoint
    end_rate = latest_strip["cash_rate"].iloc[-1]
    end_date = latest_strip["date"].iloc[-1]
    ax.annotate(
        f"{end_rate:.2f}%",
        xy=(end_date, end_rate),
        xytext=(15, 0),
        textcoords="offset points",
        fontsize=12,
        fontweight="bold",
        color="#e8731a",
        va="center",
        bbox=dict(
            boxstyle="round,pad=0.3",
            facecolor="white",
            edgecolor="#e8731a",
            linewidth=1.5,
        ),
    )

# --- Plot actual cash rate (dark teal step line) ---
ax.step(
    actual["date"],
    actual["rate"],
    where="post",
    color="#1a6e6e",  # dark teal matching IFM
    linewidth=2,
    zorder=2,
)

# =============================================================================
# 5) Styling to match IFM chart
# =============================================================================

# Axis limits
ax.set_xlim(pd.Timestamp("2015-01-01"), pd.Timestamp("2027-06-01"))
ax.set_ylim(-0.1, 5.5)

# Y-axis
ax.yaxis.set_major_locator(mticker.MultipleLocator(0.5))
ax.yaxis.set_major_formatter(mticker.FormatStrFormatter("%.1f"))
ax.set_ylabel("%pa", fontsize=12, rotation=0, labelpad=20, va="top", ha="right")
ax.yaxis.set_label_coords(-0.02, 1.02)

# X-axis — year labels only
ax.xaxis.set_major_locator(mdates.YearLocator())
ax.xaxis.set_major_formatter(mdates.DateFormatter("'%y"))

# Grid — light horizontal lines only
ax.grid(axis="y", color="#e0e0e0", linewidth=0.5)
ax.grid(axis="x", visible=False)

# Spines
ax.spines["top"].set_visible(False)
ax.spines["right"].set_visible(False)
ax.spines["left"].set_color("#cccccc")
ax.spines["bottom"].set_color("#cccccc")

# Tick styling
ax.tick_params(axis="both", which="both", length=0, labelsize=11, colors="#555555")

# Title
ax.set_title(
    "RBA Cash rate expectations*",
    fontsize=16,
    fontweight="bold",
    loc="left",
    pad=15,
    color="#2b2b2b",
)

# Annotations matching IFM style
ax.text(
    pd.Timestamp("2023-06-01"),
    5.15,
    "RBA cash rate\n& futures pricing",
    fontsize=10,
    fontstyle="italic",
    fontweight="bold",
    color="#2b2b2b",
    ha="center",
)

ax.text(
    pd.Timestamp("2018-06-01"),
    2.9,
    "*Successive 18 month\n  futures strips",
    fontsize=9,
    fontstyle="italic",
    color="#b5a99a",
    ha="center",
)

# Source & branding
ax.text(
    0.0,
    -0.08,
    "Source: ASX, Bloomberg",
    transform=ax.transAxes,
    fontsize=8,
    color="#999999",
    ha="left",
)

plt.tight_layout()
plt.savefig("docs/rba_cash_rate_expectations.png", dpi=300, bbox_inches="tight")
plt.savefig("docs/rba_cash_rate_expectations.pdf", dpi=300, bbox_inches="tight")
print("Saved to docs/rba_cash_rate_expectations.png")

File: R/ifm_cash_rate_chart.R
# ==============================================================================
# RBA Cash Rate Expectations (IFM-style spaghetti chart)
# ==============================================================================
# Purpose: Replicate the IFM Investors "RBA Cash rate expectations" chart.
#          Shows actual RBA cash rate as a dark teal step line, successive
#          ~18-month futures strips as grey/taupe lines, and the latest
#          futures strip highlighted in orange with endpoint annotation.
# ==============================================================================

ensure_packages <- function(pkgs) {
  missing_pkgs <- pkgs[!pkgs %in% rownames(installed.packages())]
  if (length(missing_pkgs) > 0) {
    install.packages(missing_pkgs, repos = "https://cloud.r-project.org")
  }
}

ensure_packages(c(
  "dplyr", "ggplot2", "here", "htmlwidgets", "lubridate", "purrr",
  "plotly", "readrba", "scales", "stringr", "tibble", "tidyr"
))

suppressPackageStartupMessages({
  library(dplyr)
  library(ggplot2)
  library(here)
  library(htmlwidgets)
  library(lubridate)
  library(purrr)
  library(plotly)
  library(readrba)
  library(scales)
  library(stringr)
  library(tibble)
  library(tidyr)
})

# ------------------------------------------------------------------------------
# 1) Load and prepare data
# ------------------------------------------------------------------------------

if (!dir.exists(here("docs", "plots", "ifm_style"))) {
  dir.create(here("docs", "plots", "ifm_style"), recursive = TRUE)
}

# Load consolidated futures data
cash_rate <- readRDS("combined_data/all_data.Rds")

# One observation per (scrape_date, contract expiry)
cash_rate_daily <- cash_rate %>%
  mutate(
    scrape_date = as.Date(scrape_time),
    expiry      = as.Date(date)
  ) %>%
  group_by(scrape_date, expiry) %>%
  slice_max(scrape_time, n = 1, with_ties = FALSE) %>%
  ungroup()

# Historical actual cash rate from RBA
actual_cash_rate <- read_rba(series_id = "FIRMMCRTD") %>%
  transmute(date, actual_rate = value) %>%
  filter(date >= as.Date("2015-01-01"))

# Extend actual rate to today so step line reaches the present
last_actual <- actual_cash_rate %>% slice_max(date, n = 1)
if (last_actual$date < Sys.Date()) {
  actual_cash_rate <- bind_rows(
    actual_cash_rate,
    tibble(date = Sys.Date(), actual_rate = last_actual$actual_rate)
  )
}

# ------------------------------------------------------------------------------
# 2) Sample monthly futures strips
# ------------------------------------------------------------------------------

# Pick one scrape per month (closest to mid-month)
sampled_scrapes <- cash_rate_daily %>%
  distinct(scrape_date) %>%
  mutate(
    ym   = floor_date(scrape_date, "month"),
    dist = abs(day(scrape_date) - 15)
  ) %>%
  group_by(ym) %>%
  slice_min(dist, n = 1, with_ties = FALSE) %>%
  ungroup() %>%
  pull(scrape_date)

# Always include the latest scrape
latest_scrape <- max(cash_rate_daily$scrape_date)
sampled_scrapes <- unique(c(sampled_scrapes, latest_scrape))

# Build strips: keep only ~18 months forward from each scrape date
strips <- cash_rate_daily %>%
  filter(scrape_date %in% sampled_scrapes) %>%
  filter(expiry <= scrape_date + months(18)) %>%
  arrange(scrape_date, expiry)

old_strips    <- strips %>% filter(scrape_date != latest_scrape)
latest_strip  <- strips %>% filter(scrape_date == latest_scrape)

# Endpoint for annotation
endpoint <- latest_strip %>% slice_max(expiry, n = 1)

# ------------------------------------------------------------------------------
# 3) Build static ggplot (IFM style)
# ------------------------------------------------------------------------------

ifm_teal   <- "#1a6e6e"
ifm_taupe  <- "#b5a99a"
ifm_orange <- "#e8731a"

p <- ggplot() +
  # Old futures strips — grey/taupe
  geom_line(
    data = old_strips,
    aes(x = expiry, y = cash_rate, group = scrape_date),
    colour    = ifm_taupe,
    linewidth = 0.4,
    alpha     = 0.55
  ) +
  # Latest futures strip — orange
  geom_line(
    data = latest_strip,
    aes(x = expiry, y = cash_rate),
    colour    = ifm_orange,
    linewidth = 1.8
  ) +
  # Actual cash rate — dark teal step line
  geom_step(
    data = actual_cash_rate,
    aes(x = date, y = actual_rate),
    colour    = ifm_teal,
    linewidth = 1.2,
    direction = "hv"
  ) +
  # Endpoint annotation
  annotate(
    "label",
    x     = endpoint$expiry,
    y     = endpoint$cash_rate,
    label = paste0(sprintf("%.2f", endpoint$cash_rate), "%"),
    colour    = ifm_orange,
    fill      = "white",
    fontface  = "bold",
    size      = 4,
    label.padding = unit(0.3, "lines"),
    label.size    = 0.8,
    hjust = -0.2
  ) +
  # Text annotations
  annotate(
    "text",
    x = as.Date("2023-06-01"), y = 5.15,
    label    = "RBA cash rate\n& futures pricing",
    fontface = "bold.italic",
    colour   = "#2b2b2b",
    size     = 3.5,
    hjust    = 0.5
  ) +
  annotate(
    "text",
    x = as.Date("2018-06-01"), y = 2.9,
    label    = "*Successive 18 month\n  futures strips",
    fontface = "italic",
    colour   = ifm_taupe,
    size     = 3.2,
    hjust    = 0.5
  ) +
  # Scales
  scale_x_date(
    limits      = c(as.Date("2015-01-01"), as.Date("2027-06-01")),
    date_labels = "'%y",
    date_breaks = "1 year",
    expand      = c(0.01, 0)
  ) +
  scale_y_continuous(
    limits = c(-0.1, 5.5),
    breaks = seq(0, 5.5, 0.5),
    labels = number_format(accuracy = 0.1)
  ) +
  labs(
    title   = "RBA Cash rate expectations*",
    y       = "%pa",
    x       = NULL,
    caption = "Source: ASX, Bloomberg"
  ) +
  theme_minimal(base_size = 12) +
  theme(
    plot.title       = element_text(face = "bold", size = 16, colour = "#2b2b2b",
                                    margin = margin(b = 10)),
    axis.title.y     = element_text(angle = 0, vjust = 1.05, hjust = 1, size = 11),
    axis.text        = element_text(colour = "#555555", size = 11),
    panel.grid.major.x = element_blank(),
    panel.grid.minor   = element_blank(),
    panel.grid.major.y = element_line(colour = "#e0e0e0", linewidth = 0.3),
    plot.caption     = element_text(colour = "#999999", size = 8,
                                    hjust = 0, margin = margin(t = 10)),
    plot.background  = element_rect(fill = "white", colour = NA),
    panel.background = element_rect(fill = "white", colour = NA)
  )

# Save static PNG
ggsave(
  filename = here("docs", "plots", "ifm_style", "rba_cash_rate_expectations.png"),
  plot     = p,
  width    = 10,
  height   = 6.5,
  dpi      = 300
)

# ------------------------------------------------------------------------------
# 4) Build interactive plotly version
# ------------------------------------------------------------------------------

# Rebuild with tooltip-friendly aesthetics
p_interactive <- ggplot() +
  geom_line(
    data = old_strips %>%
      mutate(
        tooltip_text = str_glue(
          "Strip date: {format(scrape_date, '%d %b %Y')}\n",
          "Expiry: {format(expiry, '%b %Y')}\n",
          "Implied rate: {number(cash_rate, accuracy = 0.001)}%"
        )
      ),
    aes(x = expiry, y = cash_rate, group = scrape_date, text = tooltip_text),
    colour    = ifm_taupe,
    linewidth = 0.4,
    alpha     = 0.55
  ) +
  geom_line(
    data = latest_strip %>%
      mutate(
        tooltip_text = str_glue(
          "Latest strip: {format(scrape_date, '%d %b %Y')}\n",
          "Expiry: {format(expiry, '%b %Y')}\n",
          "Implied rate: {number(cash_rate, accuracy = 0.001)}%"
        )
      ),
    aes(x = expiry, y = cash_rate, text = tooltip_text),
    colour    = ifm_orange,
    linewidth = 1.8
  ) +
  geom_step(
    data = actual_cash_rate %>%
      mutate(tooltip_text = str_glue("Actual rate: {actual_rate}%\nDate: {format(date, '%d %b %Y')}")),
    aes(x = date, y = actual_rate, text = tooltip_text),
    colour    = ifm_teal,
    linewidth = 1.2,
    direction = "hv"
  ) +
  scale_x_date(
    limits      = c(as.Date("2015-01-01"), as.Date("2027-06-01")),
    date_labels = "'%y",
    date_breaks = "1 year",
    expand      = c(0.01, 0)
  ) +
  scale_y_continuous(
    limits = c(-0.1, 5.5),
    breaks = seq(0, 5.5, 0.5),
    labels = number_format(accuracy = 0.1)
  ) +
  labs(
    title = "RBA Cash rate expectations*",
    y     = "%pa",
    x     = NULL
  ) +
  theme_minimal(base_size = 12) +
  theme(
    plot.title       = element_text(face = "bold", size = 16, colour = "#2b2b2b"),
    axis.title.y     = element_text(angle = 0, vjust = 1.05, hjust = 1),
    panel.grid.major.x = element_blank(),
    panel.grid.minor   = element_blank(),
    panel.grid.major.y = element_line(colour = "#e0e0e0", linewidth = 0.3),
    plot.background  = element_rect(fill = "white", colour = NA),
    panel.background = element_rect(fill = "white", colour = NA)
  )

p_plotly <- ggplotly(p_interactive, tooltip = "text") %>%
  layout(
    showlegend = FALSE,
    annotations = list(
      list(
        text = "Source: ASX, Bloomberg",
        x = 0, xref = "paper", xanchor = "left",
        y = -0.10, yref = "paper", yanchor = "top",
        showarrow = FALSE,
        font = list(size = 9, color = "#999999")
      )
    )
  )

saveWidget(
  widget       = p_plotly,
  file         = here("docs", "rba_cash_rate_expectations.html"),
  selfcontained = TRUE
)

cat("Saved static chart to docs/plots/ifm_style/rba_cash_rate_expectations.png\n")
cat("Saved interactive chart to docs/rba_cash_rate_expectations.html\n")
